From: Dominique Martinet <asmadeus@codewreck.org>
Subject: iov_iter: iov_folioq_get_pages: don't leave empty slot behind
Date: Mon, 11 Aug 2025 16:39:06 +0900

After advancing into a folioq it makes more sense to point to the next
slot than at the end of the current slot.  This should not be needed for
correctness, but this also happens to "fix" the 9p bug with
iterate_folioq() not copying properly.

Link: https://lkml.kernel.org/r/20250811-iot_iter_folio-v1-2-d9c223adf93c@codewreck.org
Signed-off-by: Dominique Martinet <asmadeus@codewreck.org>
Tested-by: Arnout Engelen <arnout@bzzt.net>
Acked-by: David Howells <dhowells@redhat.com>
Cc: Al Viro <viro@zeniv.linux.org.uk>
Cc: Christian Brauner <brauner@kernel.org>
Cc: Christian Theune <ct@flyingcircus.io>
Cc: Maximilian Bosch <maximilian@mbosch.me>
Cc: Ryan Lahfa <ryan@lahfa.xyz>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/iov_iter.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/lib/iov_iter.c~iov_iter-iov_folioq_get_pages-dont-leave-empty-slot-behind
+++ a/lib/iov_iter.c
@@ -1032,9 +1032,6 @@ static ssize_t iter_folioq_get_pages(str
 			maxpages--;
 		}
 
-		if (maxpages == 0 || extracted >= maxsize)
-			break;
-
 		if (iov_offset >= fsize) {
 			iov_offset = 0;
 			slot++;
@@ -1043,6 +1040,9 @@ static ssize_t iter_folioq_get_pages(str
 				slot = 0;
 			}
 		}
+
+		if (maxpages == 0 || extracted >= maxsize)
+			break;
 	}
 
 	iter->count = count;
_
